wp/lib/bap_wp/src/precondition.ml

```diff
@@ -1425,3 +1392,288 @@ let construct_pointer_constraint (l_orig : Constr.z3_expr list) (env1 : Env.t)
+
+(* Assumes the loop header branches into two blocks. *)
+let get_branch (blk : Blk.t) : (Jmp.t * Jmp.t) option =
+  match Seq.to_list (Term.enum jmp_t blk) with
+  | [jmp1; jmp2] -> Some (jmp1, jmp2)
+  | _ -> None
+
+(* Gets the BAP expression that represents the condition for going into the
+   loop. *)
+let get_branch_condition (loop : Graphs.Ir.Node.t group) (jmp1 : Jmp.t)
+    (jmp2 : Jmp.t) : Exp.t option =
+  match target jmp1, target jmp2 with
+  | Some target1, Some target2 ->
+    if in_loop loop target1 then
+      (* jmp1 goes to loop, jmp2 exits. while(E) is jmp1's condition. *)
+      Some (Jmp.cond jmp1)
+    else if in_loop loop target2 then
+      (* jmp1 exits, jmp2 goes to loop. while(E) is the negation of jmp1's
+         condition. *)
+      Some Bil.(unop not (Jmp.cond jmp1))
+    else
+      None
+  | _, _ -> None
+
+(* Gets a constraint representing the condition for a loop. *)
+let get_cond (loop : Graphs.Ir.Node.t group) (blk : Blk.t) (env : Env.t)
+  : Constr.t option =
+  let ctx = Env.get_context env in
+  let* jmp1, jmp2 = get_branch blk in
+  let* branch_cond = get_branch_condition loop jmp1 jmp2 in
+  let cond_val, _, env = exp_to_z3 branch_cond env in
+  let cond_size = BV.get_size (Expr.get_sort cond_val) in
+  let loop_guard =
+    Bool.mk_not ctx (Bool.mk_eq ctx cond_val (z3_expr_zero ctx cond_size))
+    |> Constr.mk_goal "Loop guard"
+    |> Constr.mk_constr
+  in
+  (* Compute to get the constraint at the top of the block. *)
+  let compute elts =
+    Seq.fold elts ~init:(loop_guard, env) ~f:(fun (pre, env) elt ->
+        match elt with
+        | `Def _ | `Phi _ -> visit_elt env pre elt
+        | `Jmp jmp ->
+          if (Jmp.equal jmp jmp1) || (Jmp.equal jmp jmp2) then
+            pre, env
+          else
+            visit_elt env pre elt)
+  in
+  let pre, _ = blk |> Blk.elts ~rev:true |> compute in
+  Some pre
+
+(* Gets the body of a loop without the header node containing the loop
+   condition. *)
+let get_body (graph : Graphs.Ir.t) (loop : Graphs.Ir.Node.t group) (blk : Blk.t)
+  : Graphs.Ir.t =
+  let outside_loop node =
+    let label = Graphs.Ir.Node.label node in
+    (Blk.equal label blk) || (not @@ in_loop loop (Term.tid label))
+  in
+  (* Removes the block that represents the loop header which determines
+     whether to jump to the loop or exit. *)
+  let enter_node _ node g =
+    if outside_loop node then
+      Graphs.Ir.Node.remove node g
+    else
+      g
+  in
+  (* Removes edges that exit the loop. *)
+  let enter_edge _ edge g =
+    let dst = Graphs.Ir.Edge.dst edge in
+    if outside_loop dst then
+      Graphs.Ir.Edge.remove edge g
+    else
+      g
+  in
+  Graphlib.depth_first_search (module Graphs.Ir)
+    ~enter_node ~enter_edge ~init:graph graph
+
+(* Splits the loop into the condition and the body of the loop. Assumes there
+   are no breaks in the loop. *)
+let split_loop (loop : Graphs.Ir.Node.t group) (graph : Graphs.Ir.t) (env : Env.t)
```

Comment:
Is it possible to have a dummy precondition for the loop exit to  avoid having to do this split?

Reply:
We should discuss, I'm not sure.